### ISSUE TITLE
"First frame problem".

### ISSUE THREAD

**Issue by artem-zinnatullin** [lyft/domic#14]
There is a common problem with async rendering: due to async data arrival and async switching to ui thread you're effectively skipping first frame.

In many cases it's not really a problem because it's very hard to notice one frame without data, especially if default layout state already displays something.

However in some cases like very different default state of the layout compared to the state with data (ie whole content is invisible by default, etc) it can be noticeable.

This GitHub issue tracks possible solutions and general discussion of this problem.

cc @alexjlockwood @kxfang

Also related to RxAndroid https://github.com/ReactiveX/RxAndroid/pull/228

**Comment by artem-zinnatullin** [comment]
Right now Domic renders everything by [posting to `Choreographer`](https://github.com/lyft/domic/blob/0eca77ce36480e4d2e63089f728e35a1ad83943b/domic/android/src/main/java/com/lyft/domic/android/rendering/AndroidRenderer.kt#L50) at constant time intervals. 

Changes posted to Domic's DOM do skip current frame and are rendered on the next frame which is *not a problem in general* since data arrives asynchronously in the first place but it falls into "First frame problem" case.

#### Possible solution:

Expose an API to force rendering of the current buffer ie `Renderer.renderCurrentBuffer()` (naming is not final). 

Default `Renderer` implementation, `AndroidRenderer` will render current buffer synchronously if `renderCurrentBuffer()` was called on main thread (throw exception if called on non main thread?).

Call to this method can be integrated into userland code or frameworks like [Scoop](https://github.com/lyft/scoop), Fragments, Activity, etc to render current buffer when first lifecycle method finishes thus giving user a way to solve "First frame problem"

ie:

```kotlin
class MyActivity : Activity {

    override fun onCreate(…) {
        setContentView(R.layout.myactivity)

        // Bind Domic, push some changes to render.

        renderer.renderCurrentBuffer()
    }

}
```

We could do some heuristics inside Domic to make it automatic, but that doesn't look scalable and flexible enough to be able to integrate with different frameworks like Scoop.

**Comment by artem-zinnatullin** [comment]
Please note that lots of apps already skip first frame due to `observeOn(mainThread)` applied everywhere before interaction with the view layer which posts to `Handler` thus skipping current frame

See also:

- https://github.com/JakeWharton/RxBinding/issues/28
- https://github.com/ReactiveX/RxAndroid/pull/228
- https://github.com/ReactiveX/RxAndroid/pull/416


Solving this problem is not the goal of Domic's existence, but looks like we can do something about it in more controllable way than RxAndroid or RxBinding can since we have rendering pipeline with buffering.

**Comment by Dimezis** [comment]
I'm not sure that something like `renderer.renderCurrentBuffer()` will work. Given your data sources are asynchronous and emitting from different threads, you might just not have all the data needed to render in Activity's onCreate. 
So first of all, data sources have to have a cached data and emit it synchronously on subscription, but I suppose that kinda contradicts to Domic's idea.

**Comment by artem-zinnatullin** [comment]
@Dimezis we obviously can't solve async data problem in Domic on general sense, that's not a goal and as you noted Domic is actually async-only by design

However, often times even in async systems data is cached in memory and is effectively accessed synchronously! Unless there are enforced thread jumps.

For example just yesterday @kxfang measured that about 66% of data in golden path of our app comes from memory but due to `observeOn(mainThread)` we're still skipping first frame.

I linked issues from other Rx projects because it's a known problem in async/reactive systems and there are cases when, yes, you're writing async code but it actually gets synchronously executed.

This API won't solve problem of async data arrival but it'll help get the first frame  for data cached in memory which is not bad right :)

We won't be promoting it as "The Solution" but rather like "We did what we could on our side and gave you the tools, now it's up to you :)"
